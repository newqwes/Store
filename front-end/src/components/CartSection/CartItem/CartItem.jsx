import React from 'react';
import { getOr } from 'lodash/fp';
import PropTypes from 'prop-types';

import FONT_SIZE from '../../../constants/fontSize';
import FONT_WEIGHT from '../../../constants/fontWeight';
import { ALIGN_ITEMS, JUSTIFY_CONTENT } from '../../../constants/position';

import { cartItemType } from '../../../propType';

import Flex from '../../Flex';
import Label from '../../Label';
import Button from '../../Button';
import ValueFormatter from '../../ValueFormatter';

import {
  CartItemContent,
  CartPhotoContent,
  CartDescriptionContent,
  CartPriceContent,
} from './styled';

const CartItem = ({ item, removeFromCart, addToCart }) => {
  const { photoUrl, name, description, options, unitSign, currencySign, count } = item;

  const price = getOr(0, 'price', options);
  const weight = getOr(0, 'weight', options);

  const addOne = () => addToCart(item);
  const removeOne = () => removeFromCart(item);

  return (
    <CartItemContent>
      <Flex justifyContent={JUSTIFY_CONTENT.spaceBetween} alignItems={ALIGN_ITEMS.stretch}>
        <CartPhotoContent backgroundImage={photoUrl} />
        <CartDescriptionContent>
          <Label text={name} fontSize={FONT_SIZE.medium} fontWeight={FONT_WEIGHT.bold} />
          <Label
            text={`Описание: ${description}`}
            fontSize={FONT_SIZE.small}
            fontWeight={FONT_WEIGHT.normal}
          />
        </CartDescriptionContent>
        <CartPriceContent>
          <Flex justifyContent={JUSTIFY_CONTENT.flexEnd}>
            <ValueFormatter
              value={weight}
              fontSize={FONT_SIZE.small}
              fontWeight={FONT_WEIGHT.normal}
              postfix={unitSign}
            />
          </Flex>
          <Flex justifyContent={JUSTIFY_CONTENT.flexEnd}>
            <ValueFormatter
              value={price}
              fontSize={FONT_SIZE.small}
              fontWeight={FONT_WEIGHT.normal}
              postfix={currencySign}
            />
          </Flex>
          <Flex justifyContent={JUSTIFY_CONTENT.flexEnd}>
            <Button text='-' onClick={removeOne} />
            <Label text={count} fontSize={FONT_SIZE.small} fontWeight={FONT_WEIGHT.normal} />
            <Button text='+' onClick={addOne} />
          </Flex>
        </CartPriceContent>
      </Flex>
    </CartItemContent>
  );
};

CartItem.propTypes = {
  item: cartItemType.isRequired,
  addToCart: PropTypes.func.isRequired,
  removeFromCart: PropTypes.func.isRequired,
};

export default CartItem;
