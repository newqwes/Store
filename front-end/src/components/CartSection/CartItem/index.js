export { default } from './CartItem.jsx';
