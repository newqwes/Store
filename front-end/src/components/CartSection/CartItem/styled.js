import styled, { css } from 'styled-components';
import { get } from 'lodash/fp';

export const CartItemContent = styled.div(
  ({ themeVariant }) => css`
    margin-bottom: 15px;
    background-color: white;
    padding: 10px;
    border-radius: 5px;
    box-shadow: 0 0 10px -8px ${get(['theme', 'cart', 'shadow', themeVariant])};
  `,
);

export const CartPhotoContent = styled.div`
  flex-basis: 15%;
  background: no-repeat center/80% url(${props => props.backgroundImage});
  height: 80px;
  align-self: center;
`;

export const CartDescriptionContent = styled.div`
  flex-basis: 65%;

  span {
    display: block;
    margin-top: 10px;
  }
`;

export const CartPriceContent = styled.div`
  flex-basis: 20%;

  span {
    min-width: 40px;
    text-align: center;
  }

  div {
    margin-top: 10px;
  }

  button {
    font-size: 16px;
    width: 30px;
    height: 30px;
  }
`;
