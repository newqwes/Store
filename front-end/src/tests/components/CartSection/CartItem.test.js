import React from 'react';
import { mountWithTheme } from '../../helpers';

import {
  CartItemContent,
  CartPhotoContent,
  CartDescriptionContent,
  CartPriceContent,
} from '../../../components/CartSection/CartItem/styled';

import CartItem from '../../../components/CartSection/CartItem';
import mockProductItem from '../../helpers/mocks';

const setUp = (props) =>
  mountWithTheme(
    <CartItem addToCart={() => {}} removeFromCart={() => {}} item={mockProductItem} {...props} />
  );

describe('snapshot CartItem', () => {
  test('snapshot', () => {
    const wrapper = setUp();

    expect(wrapper).toMatchSnapshot();
  });
});

describe('CartItem component', () => {
  const mockAddToCart = jest.fn();
  const mockRemoveFromCart = jest.fn();

  const wrapper = setUp({ addToCart: mockAddToCart, removeFromCart: mockRemoveFromCart });

  it(`should render CartItem with following children:
  - 1 CartPhotoContent,
  - 1 CartItemContent,
  - 1 CartDescriptionContent,
  - 1 CartPriceContent,
  - 4 Flex,
  - 3 Label,
  - 2 ValueFormatter,
  - 2 Button`, () => {
    const photoContent = wrapper.find(CartPhotoContent);
    expect(photoContent.length).toBe(1);

    const itemContent = wrapper.find(CartItemContent);
    expect(itemContent.length).toBe(1);

    const descriptionContent = wrapper.find(CartDescriptionContent);
    expect(descriptionContent.length).toBe(1);

    const priceContent = wrapper.find(CartPriceContent);
    expect(priceContent.length).toBe(1);

    const flexs = wrapper.find('Flex');
    expect(flexs.length).toBe(4);

    const labels = wrapper.find('Label');
    expect(labels.length).toBe(3);

    const valueFormatters = wrapper.find('ValueFormatter');
    expect(valueFormatters.length).toBe(2);

    const buttons = wrapper.find('Button');
    expect(buttons.length).toBe(2);
  });
});
