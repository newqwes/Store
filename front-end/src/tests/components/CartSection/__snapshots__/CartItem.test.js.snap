// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot CartItem snapshot 1`] = `
<div
  className="sc-bdfBwQ irBXUF"
>
  <div
    className="sc-eCssSg bFyEms"
    direction="row"
  >
    <div
      className="sc-gsTCUz liCrEf"
    />
    <div
      className="sc-dlfnbm rDTbG"
    >
      <span
        className="sc-jSgupP chkdAN"
        fontSize={19}
        fontWeight={500}
      >
        Pizza
      </span>
      <span
        className="sc-jSgupP hyMFLF"
        fontSize={14}
        fontWeight={400}
      >
        Описание: none description
      </span>
    </div>
    <div
      className="sc-hKgILt iFSOwF"
    >
      <div
        className="sc-eCssSg kKWOIi"
        direction="row"
      >
        <span
          className="sc-iBPRYJ dRUKeg"
          fontSize={14}
          fontWeight={400}
        >
          0
          г
        </span>
      </div>
      <div
        className="sc-eCssSg kKWOIi"
        direction="row"
      >
        <span
          className="sc-iBPRYJ dRUKeg"
          fontSize={14}
          fontWeight={400}
        >
          0
          руб.
        </span>
      </div>
      <div
        className="sc-eCssSg kKWOIi"
        direction="row"
      >
        <button
          className="sc-gKsewC gLuNHi"
          onClick={[Function]}
        >
          -
        </button>
        <span
          className="sc-jSgupP hyMFLF"
          fontSize={14}
          fontWeight={400}
        />
        <button
          className="sc-gKsewC gLuNHi"
          onClick={[Function]}
        >
          +
        </button>
      </div>
    </div>
  </div>
</div>
`;
